### ISSUE TITLE
Build failure: pythonPackages.cffi on x86_64-darwin

### ISSUE THREAD

**Issue by theCapypara** [NixOS/nixpkgs#322191]
### Steps To Reproduce

1. build *cffi* (from 24.05, but unstable is the same at the moment) on x86_64-darwin (macOS 13.6.6)

### Build log

```
building '/nix/store/1rbs81pi4axxxpj64lvc9x6vdwkp4q7i-python3.11-cffi-1.16.0.drv'...
error: builder for '/nix/store/1rbs81pi4axxxpj64lvc9x6vdwkp4q7i-python3.11-cffi-1.16.0.drv' failed with exit code 1;
              last 25 log lines:
              >      /nix/store/jnc0a7fgq1k8zgflgjkmkijghkmps6ha-python3.11-pytest-8.1.1/lib/python3.11/site-packages/_pytest/unraisableexception.py:80: PytestUnraisableExceptionWarning: Exception ignored from cffi callback <function TestNewFFI1.test_callback_decorator.<locals>.cb at 0x104d3ef20>, trying to convert the result back to C: None
              >
              >      Traceback (most recent call last):
              >          File "/private/tmp/nix-build-python3.11-cffi-1.16.0.drv-0/cffi-1.16.0/testing/cffi1/test_new_ffi_1.py", line 1341, in test_callback_decorator
              >              assert cb((1 << (sz*8-1)) - 1, -10) == 42
              >      OverflowError: int too big to convert
              >
              >          warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))
              >
              > testing/cffi1/test_recompiler.py::test_macro_var_callback
              >      /nix/store/jnc0a7fgq1k8zgflgjkmkijghkmps6ha-python3.11-pytest-8.1.1/lib/python3.11/site-packages/_pytest/unraisableexception.py:80: PytestUnraisableExceptionWarning: Exception ignored from cffi callback <function test_macro_var_callback.<locals>.get_my_value at 0x105a15580>, trying to convert the result back to C: None
              >
              >      Traceback (most recent call last):
              >          File "/nix/store/jnc0a7fgq1k8zgflgjkmkijghkmps6ha-python3.11-pytest-8.1.1/lib/python3.11/site-packages/_pytest/python_api.py", line 960, in raises
              >              func(*args[1:], **kwargs)
              >      TypeError: initializer for ctype 'int *' must be a cdata pointer, not str
              >
              >          warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))
              >
              > -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
              > =========================== short test summary info ============================
              > FAILED testing/cffi1/test_verify1.py::test_enum_usage - cffi.VerificationError: CompileError: command '/nix/store/h0a61pim6lj25qnnw...
              > FAILED testing/cffi1/test_verify1.py::test_named_pointer_as_argument - cffi.VerificationError: CompileError: command '/nix/store/h0a61pim6lj25qnnw...
              > = 2 failed, 1888 passed, 150 skipped, 4 xfailed, 11 warnings in 366.64s (0:06:06) =
```

### Additional context

The Mac is an emulated system in KVM/QEMU, in case that is somehow relevant.

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
marco@Marcos-iMac-Pro all % nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-darwin"`
 - host os: `Darwin 22.6.0, macOS 10.16`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.23.0`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by theCapypara** [comment]
I forgot to ping the maintainers. I don't seem to be able to ping a team, so I hope it's OK I ping @mweinelt, you updated it last.

**Comment by theCapypara** [comment]
Oh, I just noticed: These are the two tests fixed by https://github.com/NixOS/nixpkgs/commit/0b499eb86239cae23e52ba7c8ad2c3d30e6ed5b8

So I guess somehow this patch isn't applied here even though it should be. I will investigate.

**Comment by theCapypara** [comment]
Clang 16 promoted these to an error:
```
Wnull-pointer-subtraction
Wimplicit-function-declaration
Wno-implicit-int
Wno-incompatible-function-pointer-types
```

Not sure which of these exactly causes the failure, I have absolutely no experience in C, but disabling all of those lets the test pass. Since I'm not really confident this is even correct, I will stop here and let somebody with more experience in this take over.

**Comment by theCapypara** [comment]
I couldn't really figure out why the tests are still failing, the patches are applied. I will leave this for somebody more experienced with C and Clang to investigate.

**Comment by theCapypara** [comment]
Okay I got it.
Instead of the patch in the commit above, pulling in this commit as a patch for all compilers fixes the issue:
https://github.com/python-cffi/cffi/commit/39bdab23615a83c1001ed822f974ae52020201ba

I don't really get why the patch that disables the warning doesn't work, but this commit just fixes why the warning is even raised.
This will be included in 1.17.0.

**Comment by theCapypara** [comment]
Seems to have been fixed via #322388
